hydra-node/src/Hydra/Chain/Direct.hs

```diff
@@ -196,8 +196,15 @@ withDirectChain tracer networkId iocp socketPath keyPair party cardanoKeys point
               (submitTx queue)
       )
       ( handle onIOException $ do
+          -- NOTE: We can't re-query the time handle while the
+          -- 'chainSyncHandler' is running due to constraints. So this will use
+          -- always these initial parameters (as queried) for time conversions.
```

Comment:
Querying things from the node involves quite some `IO` functions. However, the whole purpose of the `ChainSyncHandler` abstraction was to decouple the code for testability and not tie ourselves to `IO`, but allow running the handlers in `IOSim`. Hence we only have `MonadDelay, MonadTime, etc.` constraints on them. Also, I don't think that `MonadIO IOSim` is a thing.

Then, thinking about this.. I don't think we would want to re-query stuff from the cardano-node *on each received block*?
- `SystemStart` will never change -> ez
- `EraHistory` does change / extend in-between hard-forks? -> I have no idea.

Long story short, we query on start and this likely means we won't be able to cross hard-forks with changing slot lengths because of this.. but YAGNI?

Reply:
> EraHistory does change / extend in-between hard-forks? -> I have no idea.

It does, every era adds a new portion to it. But more importantly, the last era in the history has an upper bound which is at k blocks after the tip. Which means that, it's already obsolete by the moment you've fetched it. As you time goes by, your visible future shorten (how poetic) until the moment when the upper bound of your era history is now in the past and any slot conversion will fail. 

There are two ways around it. 

1. We can [`unsafeExtendSafeZone`](https://github.com/input-output-hk/ouroboros-network/blob/c82309f403e99d916a76bb4d96d6812fb0a9db81/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Qry.hs#L442) to infinity which is basically the same as saying "we assume that all future eras will have the same protocol parameters w.r.t slot arithmetic". Possibly okay-ish if we also allow to restart a hydra-node without loosing a head (such that, in case the parameters in a next hard-fork are bound to change, we can ship an update of the hydra-node which can accommodate this).

2. We fetch a new EraHistory everytime we need to calculate time. I imagine that defining a `MonadEraHistory` wouldn't be too complicated, with an `IOSim` instance that returns an history with no upper bound for testing, and runs in plain IO for production.

Reply:
That would mean that the current interpreter will fail after using for longer than the safe zone .. should validate that, but a `MonadTimeHandle` which could be used to mock `queryTimeHandle` out for IOSim is a good idea.